File: 42841.java
import java.util.*;
class Solution {
    String[] nums = {"1","2","3","4","5","6","7","8","9"};
    Stack<String> st = new Stack();
    public int solution(int[][] baseball) {
        int answer = 0;
        
        //3자리 수를 모두 만든다.
        //{1~9까지 중복X 순서O}
        int n = nums.length;
        int r = 3;
        LinkedList<Integer> perArr = new LinkedList();
        int[] checkArr = new int[n];
        permutation(n,r,perArr,checkArr);
        while(st.isEmpty() == false){
            String str = st.peek();
            String[] num = st.pop().split("");

            boolean result = true;
            for(int i = 0 ; i < baseball.length ; i++){
                int[] item = baseball[i];
                String[] compareNum = String.valueOf(item[0]).split("");
                int strike = item[1];
                int ball = item[2];
                
                //num 과 compareNum을 비교하여 strike와 ball 의 갯수가 맞을 경우 true
                int comStrike = 0;
         
                for(int j = 0 ; j < 3 ; j++){
                    if(num[j].equals(compareNum[j])) comStrike++;
                }
                
                int comBall = 0;
                for(int j = 0 ; j < 3 ; j++){
                    if(str.contains(compareNum[j])) comBall++;
                }
                comBall = comBall - comStrike;
         
                if(ball == comBall && strike == comStrike){
                }else{
                    result = false;
                }
            
            }//loop of i
            // 
            if(result) answer++;
            
        }
        
    
        
        return answer;
    }
    public void permutation(int n, int r, LinkedList<Integer> perArr, int[] checkArr){
        
        if(r==perArr.size()){
            StringBuilder builder = new StringBuilder();
            for(int i : perArr){
                builder.append(nums[i]);
            }
            st.push(new String(builder));
            return;
        }
        for(int i = 0 ; i < n ; i++){
            if(checkArr[i] == 0){
                perArr.add(i);
                checkArr[i] = 1;
                permutation(n,r,perArr,checkArr);
                perArr.removeLast();
                checkArr[i] = 0;
            }
        }
    }
}

File: 42841.swift
import Foundation

var stack : [[Int]] = []
func solution(_ baseball:[[Int]]) -> Int {
    var answer = 0
    let arr = [1,2,3,4,5,6,7,8,9]
    let n = arr.count
    let r = 3
    let output : [Int] = Array(repeating: 0, count: r)
    let visited : [Bool] = Array(repeating: false, count: n)
    permutation(arr,output,visited,0,n,r)
    while stack.isEmpty == false {
        var compare : [Int] = stack.removeFirst()
        var flag = true
        for item in baseball{
            var strike = 0
            var ball = 0
            var num = item[0]
            var tmpStrike = item[1]
            var tmpBall = item[2]
            var numArray : [Int] = String(num).compactMap{$0.wholeNumberValue}
            for i in 0..<3 {
                if compare[i] == numArray[i] {
                    strike += 1
                }
                if compare.contains(numArray[i]){
                    ball += 1
                }
            }
            ball = ball - strike
            if strike == tmpStrike && ball == tmpBall {
                
            }else{
                // print(compare,numArray,strike,ball)
                flag = false
            }
        }//end of loop
        if flag == true {
            answer += 1
        }   
    }
  
    
    return answer
}
func permutation(_ arr : [Int],_ output : [Int],_ visited : [Bool],_ depth : Int,_ n : Int,_ r : Int){
    var arr = arr
    var output = output
    var visited = visited

    if depth == r {
        stack.append(output)
        // print(output)
        return
    }
    for i in 0..<n {
        if visited[i] != true {
            visited[i] = true
            output[depth] = arr[i]
            permutation(arr,output,visited,depth + 1,n,r)
            output[depth] = 0
            visited[i] = false
        }
    }
}
